testutils/infra/device.py

```diff
@@ -0,0 +1,217 @@
+
+        Keyword arguments:
+        file - local filename
+        local_path - local dirpath
+        remote_path - remote dirpath
+        """
+        if not fabric.api.env.host_string:
+            fabric.api.execute(
+                self.put, file, local_path, remote_path, hosts=self.host_string
+            )
+            return
+
+        _put(file, local_path, remote_path)
+
+    def ssh_is_opened(self, wait=60 * 60):
+        """Block until SSH connection is established on the device
+
+        Keyword arguments:
+        wait - Timeout (in seconds)
+        """
+        self.run("true", hide=True, wait=wait)
+
+
+class MenderDeviceGroup:
+    """Group of SSH accessible devices with Mender client
+
+    Currently, run/put methods are serialized, but they will be migrated to a parallel execution with Fabric 2
+    https://docs.fabfile.org/en/latest/api/group.html
+    """
+
+    def __init__(self, host_string_list, user="root"):
+        self._devices = []
+        for host_string in host_string_list:
+            self._devices.append(MenderDevice(host_string))
+
+    def __len__(self):
+        return len(self._devices)
+
+    def __getitem__(self, idx):
+        return self._devices[idx]
+
+    def run(self, cmd, **kw):
+        """Run command for all devices in group in a serialized manner
+
+        see MenderDevice.run
+        """
+        output_dict = dict()
+        for dev in self._devices:
+            output = dev.run(cmd, **kw)
+            output_dict[dev.host_string] = output
+        return output_dict
+
+    def ssh_is_opened(self, wait=60 * 60):
+        """Block until SSH connection is established for all devices in group in a serialized manner
+
+        see MenderDevice.ssh_is_opened
+        """
+        for dev in self._devices:
+            dev.ssh_is_opened(wait)
+
+
+def _ssh_prep_args():
+    return _ssh_prep_args_impl("ssh")
+
+
+def _scp_prep_args():
+    return _ssh_prep_args_impl("scp")
+
+
+def _ssh_prep_args_impl(tool):
+    if not fabric.api.env.host_string:
+        raise Exception("get()/put() called outside of execute()")
+
+    cmd = "%s -C -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null" % tool
+
+    host_parts = fabric.api.env.host_string.split(":")
+    host = ""
+    port = ""
+    port_flag = "-p"
+    if tool == "scp":
+        port_flag = "-P"
+    if len(host_parts) == 2:
+        host = host_parts[0]
+        port = "%s%s" % (port_flag, host_parts[1])
+    elif len(host_parts) == 1:
+        host = host_parts[0]
+        port = ""
+    else:
+        raise Exception("Malformed host string")
+
+    return (cmd, host, port)
+
+
+def _put(file, local_path=".", remote_path="."):
+    (scp, host, port) = _scp_prep_args()
+
+    fabric.api.local(
+        "%s %s %s/%s %s@%s:%s"
+        % (scp, port, local_path, file, fabric.api.env.user, host, remote_path)
+    )
+
+
+def _run(cmd, **kw):
+    if kw.get("wait") is not None:
+        wait = kw["wait"]
```

Comment:
Python scoping is weird

Reply:
> So much to like here!
> 
> Good job! 1st_place_medal raised_hands checkered_flag cherries
> 
> Only playing with some ideas, and some questions (:

Thank you. This is the most emojis I have ever received for a PR :muscle: 

And thanks for your comments. I locally fixed the small grammar stuff and I'll wait for one more review before doing the `Helpers` thing.